contrib/colab/DC-GAN.ipynb

```diff
@@ -0,0 +1,557 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "OFzyeN_GQhxk",
+        "colab": {}
+      },
+      "source": [
+        "def train_gan(rank):\n",
+        "    torch.manual_seed(1)\n",
+        "    \n",
+        "    if not xm.is_master_ordinal():\n",
+        "        # Barrier: Wait until master is done downloading\n",
+        "        xm.rendezvous('download_only_once')\n",
+        "    # Dataset\n",
+        "    data = mnist_data()\n",
+        "    if xm.is_master_ordinal():\n",
+        "        # Master is done, other workers can proceed now\n",
+        "        xm.rendezvous('download_only_once')\n",
+        "    train_sampler = torch.utils.data.distributed.DistributedSampler(\n",
+        "        data,\n",
+        "        num_replicas=xm.xrt_world_size(),\n",
+        "        rank=xm.get_ordinal(),\n",
+        "        shuffle=True)\n",
+        "    \n",
+        "\n",
+        "    # Create loader with data, so that we can iterate over it\n",
+        "    #train_loader = torch.utils.data.DataLoader(data, batch_size=100, shuffle=True)\n",
+        "    train_loader = torch.utils.data.DataLoader(\n",
+        "      data,\n",
+        "      batch_size=FLAGS['batch_size'],\n",
+        "      sampler=train_sampler,\n",
+        "      num_workers=FLAGS['num_workers'],\n",
+        "      drop_last=True)\n",
+        "\n",
+        "    # Num batches\n",
+        "    num_batches = len(train_loader)\n",
+        "    \n",
+        "    device = xm.xla_device()\n",
+        "    \n",
+        "    generator = GenerativeNet().to(device)\n",
+        "    generator.apply(init_weights)\n",
+        "\n",
+        "    discriminator = DiscriminativeNet().to(device)\n",
+        "    discriminator.apply(init_weights)\n",
+        "    \n",
+        "    # Optimizers\n",
+        "    d_optimizer = Adam(discriminator.parameters(), lr=FLAGS['disc_learning_rate'], betas=(0.5, 0.999))\n",
+        "    g_optimizer = Adam(generator.parameters(), lr=FLAGS['gen_learning_rate'], betas=(0.5, 0.999))\n",
+        "\n",
+        "    # Number of epochs\n",
+        "    num_epochs = FLAGS['num_epochs'] \n",
+        "    # Loss function\n",
+        "    loss = nn.BCELoss()\n",
+        "    \n",
+        "    num_test_samples = 24\n",
+        "    test_noise = noise(num_test_samples, device)\n",
+        "    \n",
+        "    def train_step_discriminator(optimizer, real_data, fake_data, device):\n",
+        "        # Reset gradients\n",
+        "        optimizer.zero_grad()\n",
+        "\n",
+        "        # 1. Train on Real Data\n",
+        "        prediction_real = discriminator(real_data)\n",
+        "        # Calculate error and backpropagate\n",
+        "        error_real = loss(prediction_real, real_data_target(real_data.size(0), device))\n",
```

Comment:
Just a suggestion, I think it would be more readable if instead of calling a function defined a few cells above, we had

`real_target = torch.ones(real_data.size(0), 1).to(device)`
`fake_target = torch.zeros(real_data.size(0), 1).to(device)`

Reply:
That would mean repeating those lines in the train_step_descriminator and train_step_generator?

Reply:
Please let me know if you recommend it. I can make that change.

Reply:
You can define it as above and pass it as an arg to those functions. I think it looks cleaner.